### PR TITLE
Datasets in README and tutorials

### PR DESCRIPTION
**Pull request by RdoubleA**
## Changelog
- Add supported datasets table to README
- Add tutorial on how to configure your own dataset, instruct templates, and chat formats
- Update some missing docs and API references

## Test plan
`cd docs; make html`

### CODE REVIEW

File: docs/source/examples/datasets.rst [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    #         "I am going to Paris, what should I see? [/INST] ",
+    #     ),
+    #     Message(
+    #         role="assistant",
+    #         content="Paris, the capital of France, is known for its stunning architecture..."
+    #     ),
+    # ]
+
+Note that the system message is now incorporated in the user message. If you create custom ChatFormats
+you can also add more advanced behavior.
```
Comment (rohan-varma) [inline docs/source/examples/datasets.rst:143]: We probably need a blurb in the tutorial on how to BYOD and how to read data from a remote file as well?

File: README.md [hunk 2]
```diff
@@ -45,6 +47,25 @@ The library currently supports the following models and fine-tuning methods.
 
 &nbsp;
 
+### Supported datasets
+
+We support any custom dataset that you provide through:
+
+- Hugging Face Hub
+- Remote files via HTTP URLs
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply README.md:55]: it's just downloaded from the url and loaded in memory

File: docs/source/api_ref_data.rst [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+torchtune.data
+==============
+
+.. currentmodule:: torchtune.data
+
+.. autosummary::
+    :toctree: generated/
+    :nosignatures:
+
+    AlpacaInstructTemplate
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply docs/source/api_ref_data.rst:13]: Where would that go? I've discussed some context in the tutorial, lmk if I should add more details elsewhere. And yeah, I can group this a bit better

File: docs/source/examples/datasets.rst [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+    # This is how you would configure the Alpaca dataset using the builder
+    dataset:
+      _component_: torchtune.datasets.instruct_dataset
+      source: tatsu-lab/alpaca
+      template: AlpacaInstructTemplate
+      train_on_input: True
+      max_seq_len: 512
+
+If you need to further customize your dataset, we strongly encourage you to create
```
[the remark this answers is not in the document]
Comment (RdoubleA) [reply docs/source/examples/datasets.rst:47]: You can configure custom datasets with our existing classes, so it's this whole section. This paragraph is about more specific use cases that would require creating an entire new dataset class. But I can add more details about this/make it more explicit

File: docs/source/examples/datasets.rst (hunk 1 above)
Comment (RdoubleA) [reply docs/source/examples/datasets.rst:143]: you mean in the first llm finetune tutorial?

**Review by ebsmothers — COMMENTED** [review]
So one comment on the overall structure here.. I feel things are a bit out of order. Like if I read this I first learn what InstructDatasets and ChatDatasets are as an abstraction. Then I learn how to configure them (but without understanding what e.g `AlpacaInstructTemplate` is). Finally I ostensibly learn how to build my own dataset, but it's basically the same config as the previous step, since we don't actually show how `my_dataset_builder` is defined. Then we go back to Instruct and Chat datasets for a deep-dive into formatting and templating.

Why not: 
a) Here is a canonical dataset from open source that everyone uses.
b) Oh actually I wanna do some surface-level param tweaking, see how easy it is do via config.
c) OK now I wanna modify the prompt format, how do I do that?
d) Then digress a bit into the different types of datasets we support, how they differ, what knobs you can configure on both.
e) Finally if I need something fundamentally different than what's currently supported, how do I write the class and hook it up to everything? I feel like the recently-added PreferenceDataset is actually a pretty good case study we could get into here.

Basically the idea here is to move from lowest complexity/generality to highest, and motivate the discussion with as many concrete examples as possible the entire way.

**Comment by RdoubleA** [comment]
@ebsmothers Yeah I like the proposed approach. Let me revisit this.

**Comment by RdoubleA** [comment]
Covered by #767
